plutus-ledger/src/Ledger/Address.hs

```diff
@@ -1,20 +1,87 @@
+import Ledger.Crypto (PrivateKey, PubKey (PubKey), PubKeyHash (PubKeyHash), pubKeyHash)
+import Ledger.Orphans ()
 import Ledger.Scripts (Validator, validatorHash)
-import Plutus.V1.Ledger.Address as Export
-import Plutus.V1.Ledger.Credential (Credential (..))
-import Plutus.V1.Ledger.Crypto (PubKey)
+import Plutus.V1.Ledger.Address as Export hiding (pubKeyHashAddress)
+import Plutus.V1.Ledger.Credential (Credential (PubKeyCredential, ScriptCredential), StakingCredential (StakingHash))
+import PlutusTx qualified
+import PlutusTx.Lift (makeLift)
+import PlutusTx.Prelude qualified as PlutusTx
+import Prettyprinter (Pretty)
+
+newtype PaymentPrivateKey = PaymentPrivateKey { unPaymentPrivateKey :: PrivateKey }
+
+newtype PaymentPubKey = PaymentPubKey { unPaymentPubKey :: PubKey }
+    deriving stock (Eq, Ord, Generic)
+    deriving anyclass (ToJSON, FromJSON, ToJSONKey, FromJSONKey, OpenApi.ToSchema)
+    deriving newtype (PlutusTx.Eq, PlutusTx.Ord, Serialise, PlutusTx.ToData, PlutusTx.FromData, PlutusTx.UnsafeFromData)
+    deriving (Show, Pretty) via PubKey
+makeLift ''PaymentPubKey
+
+newtype PaymentPubKeyHash = PaymentPubKeyHash { unPaymentPubKeyHash :: PubKeyHash }
+    deriving stock (Eq, Ord, Generic)
+    deriving anyclass (ToJSON, FromJSON, ToJSONKey, FromJSONKey, OpenApi.ToSchema)
+    deriving newtype (PlutusTx.Eq, PlutusTx.Ord, Serialise, Hashable, PlutusTx.ToData, PlutusTx.FromData, PlutusTx.UnsafeFromData)
+    deriving (Show, Pretty) via PubKeyHash
+makeLift ''PaymentPubKeyHash
+
+newtype StakePubKey = StakePubKey { unStakePubKey :: PubKey }
+    deriving stock (Eq, Ord, Generic)
+    deriving anyclass (ToJSON, FromJSON, ToJSONKey, FromJSONKey, OpenApi.ToSchema)
+    deriving newtype (PlutusTx.Eq, PlutusTx.Ord, Serialise, PlutusTx.ToData, PlutusTx.FromData, PlutusTx.UnsafeFromData)
+    deriving (Show, Pretty) via PubKey
+makeLift ''StakePubKey
+
+newtype StakePubKeyHash = StakePubKeyHash { unStakePubKeyHash :: PubKeyHash }
+    deriving stock (Eq, Ord, Generic)
+    deriving anyclass (ToJSON, FromJSON, ToJSONKey, FromJSONKey, OpenApi.ToSchema)
+    deriving newtype (PlutusTx.Eq, PlutusTx.Ord, Serialise, Hashable, PlutusTx.ToData, PlutusTx.FromData, PlutusTx.UnsafeFromData)
+    deriving (Show, Pretty) via PubKeyHash
+makeLift ''StakePubKeyHash
+
+{-# INLINABLE paymentPubKeyHash #-}
+paymentPubKeyHash :: PaymentPubKey -> PaymentPubKeyHash
+paymentPubKeyHash (PaymentPubKey pk) = PaymentPubKeyHash (pubKeyHash pk)
+
+{-# INLINABLE pubKeyHashAddress #-}
+-- | The address that should be targeted by a transaction output locked by the
+-- given public payment key (with it's public stake key).
+--
+-- TODO: This should be moved to Plutus.V1(or V2).Ledger.Address with the newtypes.
+pubKeyHashAddress :: PaymentPubKeyHash -> Maybe StakePubKeyHash -> Address
+pubKeyHashAddress (PaymentPubKeyHash pkh) skh =
+    Address (PubKeyCredential pkh)
+            (fmap (StakingHash . PubKeyCredential . unStakePubKeyHash) skh)
```

Comment:
@michaelpj This was copied from `plutus` because it's own `pubKeyHashAddress` would set the stake credential to `Nothing`. Is it okay if modify that function on the `plutus` side? V1 or V2?

Reply:
Yeah, fine to modify it, it's not used by the key bit.